src/hooks/useSortByTopic.ts

```diff
@@ -1,28 +1,41 @@
 import { useMemo } from 'react'
 import useSafeState from './useSafeState'
 
-interface SortTopic {
-  topic: string
-  order: boolean
+interface SortTopic<K extends string> {
+  topic: K
+  asc: boolean
 }
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-function useSortByTopic<T>(
+function useSortByTopic<T, K extends string>(
   data: T[],
-  defaultTopic: string,
-): { sortTopic: SortTopic; setSortTopic: React.Dispatch<React.SetStateAction<SortTopic>> } {
+  defaultTopic: K,
+  compareFnFactory?: (topic: K, asc: boolean) => (lhs: T, rhs: T) => number,
+): { sortedData: T[]; sortTopic: SortTopic<K>; setSortTopic: React.Dispatch<React.SetStateAction<SortTopic<K>>> } {
   // true == 'asc' && false == 'dsc'
-  const [sortTopic, setSortTopic] = useSafeState<SortTopic>({ topic: defaultTopic, order: true })
+  const [sortTopic, setSortTopic] = useSafeState<SortTopic<K>>({ topic: defaultTopic, asc: true })
+  const [sortedData, setSortedData] = useSafeState<T[]>(data)
+
+  const { asc, topic } = sortTopic
 
   useMemo(() => {
-    if (sortTopic) {
-      data.sort((lhs, rhs) =>
-        sortTopic.order ? lhs[sortTopic.topic] - rhs[sortTopic.topic] : rhs[sortTopic.topic] - lhs[sortTopic.topic],
-      )
+    if (topic) {
+      // make a copy as not to mutate
+      const dataCopy = Object.assign([], data)
```

Comment:
Why?
```suggestion
      const dataCopy = data.slice(0)
```

Reply:
is one faster than the other?

Reply:
also i don't need the 0 there i think

Reply:
`0` is there because once upon a time @alfetopito said it's faster. He even cited something.

It is faster though insignificantly, but more importantly it's a standard way to clone an array

Reply:
Look what I found
https://jsben.ch/Wncwa

Reply:
lol, that's convincing enough

Reply:
Found it! https://stackoverflow.com/a/21514254/1272513
Doesn't seem to make that big of a difference though 

![screenshot_2020-05-29_08-01-11](https://user-images.githubusercontent.com/43217/83274326-8193f800-a182-11ea-8b0e-f607ad0db3c2.png)

Reply:
@alfetopito 
in your image `.slice()` is faster than `.slice(0)` :laughing: